kart/spatial_tree.py

```diff
@@ -0,0 +1,413 @@
+
+
+def _find_geometry_column(fields):
+    for i, field in enumerate(fields):
+        if isinstance(field, Geometry):
+            return i
+    return NO_GEOMETRY_COLUMN
+
+
+def find_s2_cells(s2_coverer, geom, transforms):
+    is_point = geom.geometry_type == GeometryType.POINT
+
+    return (
+        _point_f2_cells(s2_coverer, geom, transforms)
+        if is_point
+        else _general_s2_cells(s2_coverer, geom, transforms)
+    )
+
+
+def _apply_transform(original, transform, overwrite_original=False):
+    if transform is None:
+        return original
+    result = original if overwrite_original else original.Clone()
+    result.Transform(transform)
+    return result
+
+
+def _point_f2_cells(s2_coverer, geom, transforms):
+    import pywraps2 as s2
+
+    g = gpkg_geom_to_ogr(geom)
+
+    one_transform = len(transforms) == 1
+    if not one_transform:
+        result = set()
+
+    for transform in transforms:
+        g_transformed = _apply_transform(g, transform, one_transform)
+        p = g_transformed.GetPoint()[:2]
+        s2_ll = s2.S2LatLng.FromDegrees(p[1], p[0]).Normalized()
+        s2_token = s2.S2CellId(s2_ll.ToPoint()).ToToken()
+        if one_transform:
+            return (s2_token,)
+        else:
+            result.add(s2_token)
+
+    return result
+
+
+def _general_s2_cells(s2_coverer, geom, transforms):
+    import pywraps2 as s2
+
+    e = geom_envelope(geom)
+    if e is None:
+        return ()  # Empty.
+
+    sw_src, ne_src = (e[0::2], e[1::2])
+
+    result = set()
+
+    for transform in transforms:
+        s2_ll = []
+        for p_src in (sw_src, ne_src):
+            g = ogr.Geometry(ogr.wkbPoint)
+            g.AddPoint(*p_src)
+            _apply_transform(g, transform, overwrite_original=True)
+            p_dest = g.GetPoint()[:2]
+            s2_ll.append(s2.S2LatLng.FromDegrees(p_dest[1], p_dest[0]).Normalized())
+
+        s2_llrect = s2.S2LatLngRect.FromPointPair(*s2_ll)
+        for s2_cell_id in s2_coverer.GetCovering(s2_llrect):
+            result.add(s2_cell_id.ToToken())
+
+    return result
+
+
+@add_help_subcommand
+@click.group()
+@click.pass_context
+def spatial_tree(ctx, **kwargs):
+    """
+    Commands for maintaining an S2-cell based spatial index.
+    """
+
+
+@spatial_tree.command()
+@click.option(
+    "--all",
+    "index_all_commits",
+    is_flag=True,
+    default=False,
+    help=("Index / re-index all existing commits"),
+)
+@click.option(
+    "--clear-existing",
+    is_flag=True,
+    default=False,
+    help=("Clear existing index before re-indexing"),
+)
+@click.argument(
+    "commits",
+    nargs=-1,
+)
+@click.pass_context
+def index(ctx, index_all_commits, clear_existing, commits):
+    """
+    Indexes all features added by the supplied commits and all of their ancestors.
+    To stop recursing at a particular ancestor or ancestors (eg to stop at a commit
+    that has already been indexed) prefix that commit with a caret: ^COMMIT.
```

Comment:
so if I want to run this in a post-receive hook, and someone has pushed ref `myref`, changing it from `abc123` to `def456`, I would run this like

```
kart spatial-tree index def456 ^abc123
```

That's not super intuitive to me (and I don't think `^{commit}` is an established convention?)

Would it make more sense to take a range argument?

```
kart spatial-tree index abc123..def456
```

Reply:
So actually these arguments are just forwarded verbatim to `git rev-list`. That means:
- ^X is actually an established convention (although I hadn't heard of it)
- X..Y and X...Y actually already work. I just didn't document them since they're syntactic sugar for some X, Y, ^Z way of writing the some specification. 

The X, Y, ^Z way of specifying commits is more expressive since if you have new branch heads N1, N2, N3, N4, and old branch heads O1, O2 that you already indexed last time, you can just index: `N1 N2 N3 N4 ^O1 ^O2` without worrying which old branch head(s) are ancestors of which new branch heads (it might be true that all are ancestors of all).

Reply:
Maybe change the docstring to avoid using that and just say you can pass it anything `git rev-list` accepts